Giulio benedini/csv logger

GiulioIlBen opened the pull request:
This is a simple csv logger. 

It has been integrated in the JobManager, such that it gives you a simple summary of your finished jobs.

Commit:
csv logger added. Now you get a csv summary in the jobmanager by default with the FINAL status of your jobs

Commit:
solved typechecking issue and improved test

Commit:
solved unused imports

dormrod commented on core/jobmanager.py:
```diff
@@ -42,7 +43,7 @@ class JobManager:
     """
 
-    def __init__(self, settings, path=None, folder=None, use_existing_folder=False):
+    def __init__(self, settings, path=None, folder=None, use_existing_folder=False, csv_logger: LoggerCSV = None):
```
I agree that each job manager should have a separate log file. 

I find it odd that the PLAMS method `log` always writes to the same `default_jobmanager` log file, even when the job might be using a different job manager.

Does this need to be an argument though? Did you ever find it useful to pass in an existing csv logger?

GiulioIlBen replied:
This gives you flexibility, because ideally if you want to log all the jobs in the `default_jobmanager` you can in principle provide that one as argument and therefore centralize the jobs logs in one file.

dormrod commented on core/csv_logger.py:
```diff
@@ -0,0 +1,127 @@
+        self.log_path = Path(log_path)
+        self.stdout_level = stdout_level
+        self.date = date
+        self.time = time
```
For consistency, these should come from the global config object. 

So `config.log.date`, `config.log.time` etc. 

See: https://github.com/SCM-NV/PLAMS/blob/trunk/core/settings.py#L544

dormrod commented on core/csv_logger.py:
```diff
@@ -0,0 +1,127 @@
+        :param level: 1: important; 3: normal; 5: verbose; 7: debug, defaults to 0
+        :type level: int, optional
+        """
+        if not self.init:
```
Is this needed here? For the main `log` function, it is checking whether init has been called (and therefore the default job manager set up) but here we don't need to wait for this, or anything else. The path has been set in the `__init__` function

GiulioIlBen replied:
true. It is a remnant due to historical reasons (at the beginning `LoggerCSV` was a function like the `log` one).

dormrod commented on core/csv_logger.py:
```diff
@@ -0,0 +1,127 @@
+            row.append(message.get(key))
+        extra_headers_needed = new_headers_needed and (current_headers != [])
+
+        if extra_headers_needed:
+            raise NotImplementedError(
+                f"found new headers! {current_headers=} BUT {headers=}, this option is not currently supported"
+            )
```
Right, because otherwise you would have to re-write the file with the missing `,,,`...

Could be an argument for json lines as you mentioned, as this would be more flexible, but might be less easier for users to interact with than a csv file.


For csv though, should the headers be inferred, or just set.

As in, should this be derived class called `JobCSVLogger` and you just call `log(job)` 🤔

dormrod commented on core/csv_logger.py:
```diff
@@ -0,0 +1,127 @@
+                headers.insert(1 if self.date else 0, "Time")
+
+        # Log to file
+        with _filelock:
+            with open(self.log_path, mode="a", newline="") as file:
+                csv_writer = csv.writer(file)
+                if not current_headers:
+                    csv_writer.writerow(headers)
+                csv_writer.writerow(row)
```
I would like to take the opportunity to move to the standard python logger, instead of handling the files/locking etc. ourselves. 

I will try for the default case `log` method and let you know how it goes...

dormrod commented on unit_tests/test_csv_logger.py:
```diff
@@ -0,0 +1,22 @@
+
+
+def test_csv_logger():
+    csv_logger = LoggerCSV(log_path=os.getcwd() + "/logfile.csv")
```
Can use a temp file to avoid having to clean it up after the test:
```
with tempfile.NamedTemporaryFile() as temp_log_file:
    csv_logger = LoggerCSV(log_path=temp_log_file.name)
```

dormrod commented on core/csv_logger.py:
```diff
@@ -0,0 +1,127 @@
+        extra_headers_needed = new_headers_needed and (current_headers != [])
+
+        if extra_headers_needed:
+            raise NotImplementedError(
```
Maybe a value error?

dormrod commented:
Really nice idea and will make people a lot happier I think!

dormrod committed:
Add test for log function SCMSUITE-9740 SO107

dormrod committed:
Add logger class and unit tests SCMSUITE-9740 SO107

dormrod committed:
Update usages of temp file in tests for windows SCMSUITE-9740 SO107

dormrod committed:
Add LoggerManager SCMSUITE-9740 SO107

dormrod committed:
Add lock into logger SCMSUITE-9740 SO107

dormrod committed:
Rename to LogManager SCMSUITE-9740 SO107

dormrod committed:
Update logger test names SCMSUITE-9740 SO107

dormrod committed:
Add get_logger method SCMSUITE-9740 SO107

Commit:
Merge remote-tracking branch 'origin/DavidOrmrodMorley/default_logger' into GiulioBenedini/csv_logger

Commit:
SO-- logger abstraction and csv logger

created the Logger(ABC)
renamed the Logger to StrLogger(Logger)
added csv formatter and CSVLogger(Logger)
which basically configure the CSVFormatter

you can control which logging format via the format input for:
get_logger(name: str, format: Optional[Literal["csv",  "str"]] = "str")

Commit:
SO-- implemented a JobCSVFormatter

added a JobCSVFormatter(CSVFormatter) for logging job basic information

Commit:
SO-- bug fix for logging 28-self._stdout_handler.level
and csv logger in the jobmanager

having the same formatter for two handler is not ideal for a csv logger.
For csv formatter implemented there is an initialization step where you write down the headers, this in my opinion is best to have implemented in the formatter itself. But if both handlers have the same formatter the initialization step is called only once. therefore the Logger(ABC) now has _file_formatter and _stdout_formatter attributes

Commit:
SO-- added DictFormatter base class for both JSONFormatter and CSVFormatter.

the JSONFormatter it was easy to implement so preferred to do so but maybe not much useful for now.

Commit:
SO-- defaults changed and one test for csv

Commit:
flake bug fix

Commit:
flake bug fix and typing improvement

Commit:
SO-- added a notebook example

Commit:
Merge branch 'GiulioBenedini/csv_logger' of https://github.com/SCM-NV/PLAMS into GiulioBenedini/csv_logger

dormrod committed:
Merge branch 'trunk' into GiulioBenedini/csv_logger

dormrod committed:
Remove unused old csv logger implementation SO107

dormrod committed:
Refactor CSV logger and add tests

dormrod committed:
Update logging example

dormrod committed:
Fix unit test and flakes issues

dormrod reviewed (commented):
Summary of my changes:

* Removed old implementation
* Polished/improved the logger base class and simplified the CSV logger
* Updated the example with some explanatory text
* Added unit tests

dormrod committed:
Move job logging into job runner

dormrod committed:
Unit test and lint fixes

dormrod committed:
Update examples based

GiulioIlBen reviewed (commented):
The code looks good to me. I think we can merge it.
